### ISSUE TITLE
AVIATOR

### ISSUE THREAD

**Issue by Wilen047** [Wilen047/AVIATOR-PREDICTION-APK#12]
# Issue #12: Model Evaluation, Validation & Performance Analysis

**Status:** Open

**Priority:** P1

**Labels:** `machine-learning`, `evaluation`, `validation`, `model-performance`

## Description

Evaluate the machine-learning models developed in Issue #11 using unseen historical data and time-aware validation.

The goal is to determine whether the models provide useful information beyond simple baseline methods and whether they generalize to unseen data.

Model performance must be reported honestly. The application must not claim that a model can guarantee or reliably predict the outcome of a random Aviator round.

## Objectives

* Evaluate trained ML models.
* Compare models against baseline methods.
* Perform time-based validation.
* Measure regression/classification performance.
* Detect overfitting.
* Analyze errors.
* Generate performance reports.
* Select the best validated model.
* Document model limitations.

## Evaluation Dataset

The final test dataset must remain unseen during:

* Feature selection.
* Model training.
* Hyperparameter tuning.
* Model selection.

The test data should represent a later chronological period than the training data.

```text
Historical Data
       ↓
Training Data
       ↓
Validation Data
       ↓
Final Test Data
```

### Regression Models

Calculate:

* MAE
* RMSE
* R²
* Median Absolute Error

Example:

```text
Model: Random Forest

MAE: 0.XX
RMSE: X.XX
R²: 0.XX
```

### Classification Models

Calculate:

* Accuracy
* Precision
* Recall
* F1-score
* Confusion Matrix
* Log Loss

Use macro/weighted metrics where class imbalance exists.

## Baseline Comparison

Compare the ML model against simple baselines.

Example:

```text
Model              MAE
--------------------------------
Mean Baseline      X.XX
Median Baseline    X.XX
Previous Value     X.XX
ML Model           X.XX
```

For classification:

```text
Model              Accuracy
--------------------------------
Most Frequent      XX%
ML Model           XX%
```

A model should not be considered successful simply because it has a high score. It must demonstrate meaningful improvement over appropriate baselines on unseen chronological data.

## Time-Based Validation

Use chronological validation.

Example:

```text
Fold 1
Train → Period 1
Validate → Period 2

Fold 2
Train → Period 1 + 2
Validate → Period 3

Fold 3
Train → Period 1 + 2 + 3
Validate → Period 4
```

This helps determine whether performance remains stable over time.

## Overfitting Analysis

Compare:

```text
Training Performance
vs
Validation Performance
vs
Test Performance
```

Example:

```text
Training Score  → Very High
Validation Score → Medium
Test Score       → Low
```

A large performance gap may indicate overfitting.

## Error Analysis

For regression models, analyze:

* Prediction error distribution.
* MAE by multiplier range.
* RMSE by multiplier range.
* Large-error cases.
* Performance across different time periods.

For classification models, analyze:

* Confusion matrix.
* Per-class precision.
* Per-class recall.
* Misclassified classes.
* Class imbalance.

## Performance by Multiplier Range

Evaluate performance across configurable ranges:

```text
1.00x – 1.49x
1.50x – 1.99x
2.00x – 2.99x
3.00x – 4.99x
5.00x – 9.99x
10.00x+
```

This helps identify where the model performs well or poorly.

## Stability Analysis

Evaluate whether model performance changes across different historical periods.

Example:

```text
Period 1 → MAE: X.XX
Period 2 → MAE: X.XX
Period 3 → MAE: X.XX
Period 4 → MAE: X.XX
```

If performance changes significantly, the model should not be assumed to generalize consistently.

## Calibration

For classification models that output probabilities, evaluate whether predicted probabilities correspond reasonably to observed frequencies.

Possible tools:

* Reliability diagrams.
* Brier score.
* Calibration curves.

Probability outputs must not be presented as guaranteed chances of winning.

## Evaluation Report

Generate a structured evaluation report.

Example:

```json
{
  "model_name": "AviatorHistoricalModel",
  "model_version": "1.0.0",
  "test_period": {
    "start": "2026-08-01",
    "end": "2026-08-19"
  },
  "metrics": {
    "mae": 0.0,
    "rmse": 0.0,
    "r2": 0.0
  },
  "baseline_comparison": {
    "baseline_mae": 0.0,
    "model_mae": 0.0
  },
  "status": "evaluated"
}
```

Actual values should be generated from the real test results.

## Model Selection Rules

Define objective selection criteria before choosing the final model.

For regression:

* Lower MAE is better.
* Lower RMSE is better.
* Performance must be stable across time.
* Model should outperform appropriate baselines.

For classification:

* Higher F1-score is better.
* Performance should be reasonable across classes.
* Model should outperform the baseline.
* Calibration should be checked when probabilities are displayed.

## Model Approval

Only models that meet predefined evaluation requirements should be marked:

```text
approved
```

Models that fail evaluation should be marked:

```text
rejected
```

or:

```text
experimental
```

## API Endpoint

Create an administrator-only endpoint for retrieving evaluation results:

```text
GET /api/v1/ml/evaluation/{model_id}
```

Optional endpoint for triggering evaluation:

```text
POST /api/v1/ml/evaluate/{model_id}
```

Long-running evaluations should run as background jobs.

## Implementation Structure

Suggested structure:

```text
app/
├── ml/
│   ├── evaluation/
│   │   ├── metrics.py
│   │   ├── validation.py
│   │   ├── error_analysis.py
│   │   ├── calibration.py
│   │   └── report.py
│   │
│   └── model_registry.py
│
├── schemas/
│   └── evaluation.py
│
└── routers/
    └── ml.py
```

## Testing Requirements

* [ ] Regression metrics are tested.
* [ ] Classification metrics are tested.
* [ ] Baseline comparison is tested.
* [ ] Time-based validation is tested.
* [ ] Overfitting checks are implemented.
* [ ] Error analysis works.
* [ ] Performance by multiplier range works.
* [ ] Stability analysis works.
* [ ] Calibration analysis works where applicable.
* [ ] Evaluation reports are generated correctly.
* [ ] Model approval/rejection logic is tested.

## Acceptance Criteria

* [ ] Test dataset remains unseen during model development.
* [ ] Time-based evaluation is implemented.
* [ ] Baseline comparison is completed.
* [ ] Appropriate metrics are calculated.
* [ ] Error analysis is completed.
* [ ] Overfitting is assessed.
* [ ] Model stability is assessed.
* [ ] Performance across multiplier ranges is analyzed.
* [ ] Evaluation report is generated.
* [ ] Model version and evaluation results are stored.
* [ ] Final model selection is documented.
* [ ] Model limitations are documented.
* [ ] Evaluation tests pass.

## Expected Outcome

A complete model evaluation system that determines whether the trained models generalize to unseen historical data and whether their performance is meaningfully better than simple baselines.

The result of this issue should determine whether a model is suitable for use as an **experimental analytical estimate** in the APK.

## Next Issue

**Issue #13: Build Prediction/Analytics API**
